stackslib/src/chainstate/nakamoto/coordinator/mod.rs

```diff
@@ -0,0 +1,844 @@
+/// block-commit validation logic).
+///
+/// If this method returns None, the caller should try again when there are more Stacks blocks.  In
+/// Nakamoto, every reward cycle _must_ have a PoX anchor block; otherwise, the chain halts.
+///
+/// N.B. this method assumes that the prepare phase is comprised _solely_ of Nakamoto tenures.  It
+/// will not work if any of the prepare-phase tenures are from epoch 2.x.
+///
+/// Returns Ok(Some(reward-cycle-info)) if we found the first sortition in the prepare phase.
+/// Returns Ok(None) if we're still waiting for the PoX anchor block sortition
+/// Returns Err(Error::NotInPreparePhase) if `burn_height` is not in the prepare phase
+/// Returns Err(Error::RewardCycleAlreadyProcessed) if the reward set for this reward cycle has
+/// already been processed.
+pub fn get_nakamoto_reward_cycle_info<U: RewardSetProvider>(
+    burn_height: u64,
+    sortition_tip: &SortitionId,
+    burnchain: &Burnchain,
+    chain_state: &mut StacksChainState,
+    sort_db: &mut SortitionDB,
+    provider: &U,
+) -> Result<Option<RewardCycleInfo>, Error> {
+    let epoch_at_height = SortitionDB::get_stacks_epoch(sort_db.conn(), burn_height)?
+        .expect(&format!(
+            "FATAL: no epoch defined for burn height {}",
+            burn_height
+        ))
+        .epoch_id;
+
+    assert!(
+        epoch_at_height >= StacksEpochId::Epoch30,
+        "FATAL: called a nakamoto function outside of epoch 3"
+    );
+
+    if !burnchain.is_in_prepare_phase(burn_height) {
+        return Err(Error::NotInPreparePhase);
+    }
+
+    // calculating the reward set for the _next_ reward cycle
+    let reward_cycle = burnchain
+        .block_height_to_reward_cycle(burn_height)
+        .expect("FATAL: no reward cycle for burn height")
+        + 1;
+
+    debug!("Processing reward set for Nakamoto reward cycle";
+          "burn_height" => burn_height,
+          "reward_cycle" => reward_cycle,
+          "reward_cycle_length" => burnchain.pox_constants.reward_cycle_length,
+          "prepare_phase_length" => burnchain.pox_constants.prepare_length);
+
+    // find the last tenure-start Stacks block processed in the preceeding prepare phase
+    // (i.e. the first block in the tenure of the parent of the first Stacks block processed in the prepare phase).
+    // Note that we may not have processed it yet.  But, if we do find it, then it's
+    // unique (and since Nakamoto Stacks blocks are processed in order, the anchor block
+    // cannot change later).
+    let prepare_phase_sortitions =
+        find_prepare_phase_sortitions(sort_db, burnchain, sortition_tip)?;
+
+    // did we already calculate the reward cycle info?  If so, then return it.
+    let first_sortition_id = if let Some(first_sn) = prepare_phase_sortitions.first() {
+        if let Some(persisted_reward_cycle_info) =
+            SortitionDB::get_preprocessed_reward_set(sort_db.conn(), &first_sn.sortition_id)?
```

Comment:
Is it possible that a reward set is preprocessed BEFORE DKG is set and then this call occurs AFTER DKG is set, but the preprocessed reward set is not updated correctly to include this? (commenting mostly for my PR built ontop of this one)

Reply:
Yes, that will need to be addressed in your PR.  The reward set will need to be updated with the aggregate public key if it has since become available.  Perhaps the aggregate public key could be stored in a separate column in the same table? Then you don't have to load the entire reward set to check each time.